### PR TITLE
new requisite form: onfail_all

### LINKED ISSUE #49382: onfail requisite always triggers when combined with other requisites
### Description of Issue/Question
If using
```
onfail:
  - id: whatever
require:
  - service: httpd
```
No matter what happens with the SLS/state it's "watching", it'll always trigger as if it failed even if all the require, require_in, etc. are successful.

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
Salt Version:
           Salt: 2017.7.7
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: unknown
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: 1.6.1
       M2Crypto: 0.21.1
           Mako: 0.8.1
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: 1.2.5
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.6.1
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 4.9.123-1.el7.centos.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core

### PR DESCRIPTION
**Pull request by mattp-**
### What does this PR do?
when https://github.com/saltstack/salt/issues/22370 was introduced it
actually removed functionality, as sometimes it desirable to onfail if
ALL of the listed requisites are failure. This adds back an 'onfail_all'
form that allows for this behavior (it would make more sense to have
onfail default to OR logic, but that would break backcompat, so we add
a new form instead).

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/47015

I was in the neighborhood so also fixed https://github.com/saltstack/salt/issues/49382 and https://github.com/saltstack/salt/issues/49120

### Tests written?
Yes

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by mattp-** [comment]
although the new feature isn't really necessary for backport, the bugfixes might be worthwhile.

**Comment by max-arnold** [comment]
Typo (should be `onfail_all`): 

```
Added new `onfail_any` requisite form to allow for AND logic when adding
```

**Comment by mattp-** [comment]
@max-arnold woops thanks. fixed

File: salt/state.py [hunk 1]
```diff
@@ -2301,6 +2302,8 @@ def check_requisite(self, low, running, chunks, pre=False):
             present = True
         if 'onfail_any' in low:
             present = True
+        if 'onfail_all' in low:
+            present = True
```
[the remark this answers is not in the document]
Comment (cachedout) [reply salt/state.py]: Would be fine to fix later on. I'll resolve this for now.

File: doc/ref/states/requisites.rst [hunk 2]
```diff
@@ -155,6 +155,10 @@ Several requisite types have a corresponding `requisite_any <requisites-any>` fo
 * ``onchanges_any``
 * ``onfail_any``
 
+Lastly, onfail has one special ``onfail_all`` form to account for when `AND`
+logic is desired instead of the default `OR` logic of onfail/onfail_any (which
+are equivalent).
+
```
[the remark this answers is not in the document]
Comment (mattp-) [reply doc/ref/states/requisites.rst]: added some docs and a hopefully not too silly usage example

[the remark this answers is not in the document]
Comment (cachedout) [reply doc/ref/states/requisites.rst]: It looks great! Thanks.

**Review by cachedout — APPROVED** [review]

**Comment by cachedout** [comment]
It's odd that a Windows machine failed a state compiler test here. I'm going to update the branch and re-run the tests to see what happens.

File: tests/integration/modules/test_state.py [hunk 3]
```diff
@@ -1913,6 +1994,8 @@ def test_state_sls_id_test(self):
         testfile = os.path.join(TMP, 'testfile')
         comment = 'The file {0} is set to be changed'.format(testfile)
         ret = self.run_function('state.sls', ['core'])
+        log.error("HERE")
```
Comment (dwoz) [inline tests/integration/modules/test_state.py]: @mattp- Looks like there is some leftover debug cruft here.

Comment (mattp-) [reply tests/integration/modules/test_state.py]: actually put that there on purpose :) windows keeps failing this in jenkins and I don't know why

**Comment by mattp-** [comment]
@cachedout after pulling in @dwoz fixes for the existing windows sls_id tests, the py3 win2016 run is now clean but some others popped up but only on the py2 win run (which were passing before). So this should be safe to merge / unrelated to my changes.

**Pull request merged**